### PR TITLE
WIP : Initial changes for bundle export

### PR DESCRIPTION
**Pull request by vinu2003**
## Please provide the following details to expedite Pull Request review:
*Please replace with a description about why this change is needed?*
This PR addresses code changes to support "export bundle" functionality in juju.
Currently this feature exists currently in juju-gui. 
This export the current model into a bundle in yaml format which can then be used to deploy locally.

## QA steps
*Please replace with how we can verify that the change works?*
Yet to implement the CLI part. The idea is to implement a new command as below
'juju export-bundle [modelname]
default model or the model which is currently chosen is picked up if the model name is not specified.

## Documentation changes
*Please replace with any notes about how it affects current user workflow? CLI? API?* 
** to be filled in later **
## Bug reference
*Please add a link to any bugs that this change is related to.*
N/A

### CODE REVIEW

File: apiserver/facades/client/bundle/bundle.go [hunk 1]
```diff
@@ -91,3 +99,78 @@ func (b *bundleAPI) GetChanges(args params.BundleChangesParams) (params.BundleCh
 	}
 	return results, nil
 }
+
+// ExportBundle exports the model configuration specified in the argument as bundle.
+func (b bundleAPI) ExportBundle(args params.ExportBundleParams) (params.BundleData, error) {
```
Comment (wallyworld) [inline apiserver/facades/client/bundle/bundle.go]: The model to export comes from the facade initialisation.
The NewFacade() method which currently ignores State needs to be changed.
The New method should also be converted to use a facade context.

File: apiserver/facades/client/bundle/bundle.go [hunk 2]
```diff
@@ -91,3 +99,78 @@ func (b *bundleAPI) GetChanges(args params.BundleChangesParams) (params.BundleCh
...
+	result := params.BundleData{}
+
+	tag, err := names.ParseModelTag(args.ModelTag)
+	if err != nil {
+		return result, errors.Annotate(err, "model-tag cannot be parsed.")
+	}
+
+	st, releaseSt, err := b.state.GetBackend(tag.Id())
```
Comment (wallyworld) [inline apiserver/facades/client/bundle/bundle.go]: The backend comes from the facade state as passed in to the New() method

File: apiserver/facades/client/bundle/bundle.go [hunk 3]
```diff
@@ -91,3 +99,78 @@ func (b *bundleAPI) GetChanges(args params.BundleChangesParams) (params.BundleCh
...
+	if err != nil {
+		return result, errors.Trace(err)
+	}
+	defer releaseSt()
+
+	var model description.Model
```
Comment (wallyworld) [inline apiserver/facades/client/bundle/bundle.go]: No need to declare a var here, just use direct assignment

File: apiserver/facades/client/bundle/bundle.go [hunk 4]
```diff
@@ -91,3 +99,78 @@ func (b *bundleAPI) GetChanges(args params.BundleChangesParams) (params.BundleCh
...
+	model, err = st.Export()
+	if err != nil {
+		return result, errors.Trace(err)
+	}
+
+	// Fill in the BundleApplicationSpec of BundleData.
+	app := make(map[string]*params.BundleApplicationSpec)
```
Comment (wallyworld) [inline apiserver/facades/client/bundle/bundle.go]: We probably don't need all these new structs. The end result will just be bundle yaml. That can all be generated here server side by populating the charm.BundleData struct using the model data from Export(). The result of the api call will be a stringresult with the bundleyaml as the string value

**Comment by vinu2003** [comment]
!!build!!

File: apiserver/facades/client/bundle/bundle.go [hunk 5]
```diff
@@ -8,47 +8,52 @@ import (
 	"strings"
 
 	"github.com/juju/bundlechanges"
+	"github.com/juju/description"
 	"github.com/juju/errors"
 	"github.com/juju/loggo"
 	"gopkg.in/juju/charm.v6"
+	names "gopkg.in/juju/names.v2"
```
Comment (wallyworld) [inline apiserver/facades/client/bundle/bundle.go]: don't need alias

File: apiserver/facades/client/bundle/bundle.go [hunk 6]
```diff
@@ -8,47 +8,52 @@ import (
...
 	"github.com/juju/juju/constraints"
 	"github.com/juju/juju/core/devices"
-	"github.com/juju/juju/state"
 	"github.com/juju/juju/storage"
 )
 
-// NewFacade provides the required signature for facade registration.
-func NewFacade(_ *state.State, _ facade.Resources, auth facade.Authorizer) (Bundle, error) {
-	return NewBundle(auth)
+type Facade struct {
```
Comment (wallyworld) [inline apiserver/facades/client/bundle/bundle.go]: No need for a new struct, just use the existing bundleAPI struct

File: apiserver/facades/client/bundle/bundle_test.go [hunk 7]
```diff
@@ -8,34 +8,49 @@ import (
 	gc "gopkg.in/check.v1"
 	"gopkg.in/juju/names.v2"
 
+	"github.com/juju/description"
```
Comment (wallyworld) [inline apiserver/facades/client/bundle/bundle_test.go]: move up to other importblock

File: apiserver/facades/client/bundle/bundle_test.go [hunk 8]
```diff
@@ -8,34 +8,49 @@ import (
...
 var _ = gc.Suite(&bundleSuite{})
 
 func (s *bundleSuite) SetUpTest(c *gc.C) {
 	s.BaseSuite.SetUpTest(c)
-	auth := apiservertesting.FakeAuthorizer{
+	/*s.auth = apiservertesting.FakeAuthorizer{
 		Tag: names.NewUserTag("who"),
+	}*/
+	s.auth = &apiservertesting.FakeAuthorizer{
+		Tag: names.NewApplicationTag("gitlab"),
```
Comment (wallyworld) [inline apiserver/facades/client/bundle/bundle_test.go]: This is a client facade so the auth tag needs to be a user

File: apiserver/facades/client/bundle/mock_test.go [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+type mockState struct {
+	testing.Stub
+	bundle.Backend
+	entities map[string]state.Entity
+	app      mockApplication
+	unit     mockUnit
+}
+
+func newMockState() *mockState {
+	unitsChanges := make(chan []string, 1)
```
Comment (wallyworld) [inline apiserver/facades/client/bundle/mock_test.go]: No need for any of these watchers etc - this mock state only needs to do one thing - provide the Export() method.

File: apiserver/facades/client/bundle/mock_test.go [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+		unit: mockUnit{
+			life: state.Dying,
+		},
+	}
+	st.entities[st.app.Tag().String()] = &st.app
+	st.entities[st.unit.Tag().String()] = &st.unit
+	return st
+}
+
+type mockApplication struct {
```
Comment (wallyworld) [inline apiserver/facades/client/bundle/mock_test.go]: No need for any of this stuff apart from a simple mockState

File: apiserver/facades/client/bundle/bundle.go (hunk 5 above)
Comment (vinu2003) [reply apiserver/facades/client/bundle/bundle.go]: done

File: apiserver/facades/client/bundle/bundle.go (hunk 6 above)
Comment (vinu2003) [reply apiserver/facades/client/bundle/bundle.go]: done

File: apiserver/facades/client/bundle/bundle_test.go (hunk 7 above)
Comment (vinu2003) [reply apiserver/facades/client/bundle/bundle_test.go]: done

File: apiserver/facades/client/bundle/bundle_test.go (hunk 8 above)
Comment (vinu2003) [reply apiserver/facades/client/bundle/bundle_test.go]: ok

File: apiserver/facades/client/bundle/mock_test.go (hunk 9 above)
Comment (vinu2003) [reply apiserver/facades/client/bundle/mock_test.go]: done

File: apiserver/facades/client/bundle/mock_test.go (hunk 10 above)
Comment (vinu2003) [reply apiserver/facades/client/bundle/mock_test.go]: done

File: apiserver/facades/client/bundle/bundle.go [hunk 11]
```diff
@@ -8,47 +8,52 @@ import (
...
 	"github.com/juju/juju/constraints"
 	"github.com/juju/juju/core/devices"
-	"github.com/juju/juju/state"
 	"github.com/juju/juju/storage"
 )
 
-// NewFacade provides the required signature for facade registration.
-func NewFacade(_ *state.State, _ facade.Resources, auth facade.Authorizer) (Bundle, error) {
-	return NewBundle(auth)
+type BundleAPI struct {
```
Comment (wallyworld) [inline apiserver/facades/client/bundle/bundle.go]: There was a bundleAPI struct defined on line 46.
Please copy the comment from that struct

File: apiserver/facades/client/bundle/bundle.go [hunk 12]
```diff
@@ -96,3 +101,20 @@ func (b *bundleAPI) GetChanges(args params.BundleChangesParams) (params.BundleCh
 	}
 	return results, nil
 }
+
+// ExportBundle exports the current model configuration as bundle.
+func (b *BundleAPI) ExportBundle() (params.StringResult, error) {
+	model, err := b.backend.Export()
```
Comment (wallyworld) [inline apiserver/facades/client/bundle/bundle.go]: There needs to be a permission check that the user has read access to the model.

File: apiserver/facades/client/bundle/mock_test.go [hunk 13]
```diff
@@ -0,0 +1,31 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package bundle_test
+
+import (
+	"github.com/juju/juju/apiserver/facades/client/bundle"
+	"github.com/juju/juju/apiserver/params"
+	"github.com/juju/testing"
```
Comment (wallyworld) [inline apiserver/facades/client/bundle/mock_test.go]: fix import blocks

File: apiserver/facades/client/bundle/mock_test.go [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+)
+
+type mockState struct {
+	testing.Stub
+	bundle.Backend
+	str params.StringResult
+}
+
+func (m *mockState) ExportBundle() (params.StringResult, error) {
```
Comment (wallyworld) [inline apiserver/facades/client/bundle/mock_test.go]: The method we need here is the state method Export() - that is what we are mocking.
The idea is that the facade calls a state object to do stuff. In production it is the real state. In tests, it is a mock which must have the same interface as the real state. So mockState needs to implement migration.StateExporter. 
Also, don't embed Backend interface in mockState as that just masks the fact that it hasn't been implemented properly.
You just want
type mockState struct {
testing.Stub
}

File: apiserver/facades/client/bundle/mock_test.go [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+	m.MethodCall(m, "ExportBundle")
+	if err := m.NextErr(); err != nil {
+		return params.StringResult{}, err
+	}
+
+	str := m.str
+	return str, nil
+}
+
+func newMockState() *mockState {
```
Comment (wallyworld) [inline apiserver/facades/client/bundle/mock_test.go]: No need for this, just do &mockState{} in the test set up

File: apiserver/params/params.go [hunk 16]
```diff
@@ -1125,6 +1125,122 @@ type BundleChange struct {
 	Requires []string `json:"requires"`
 }
 
+// ExportBundleParams holds parameters for the ExportBundle call.
```
Comment (wallyworld) [inline apiserver/params/params.go]: All this can be deleted

File: apiserver/facades/client/client/bundles.go [hunk 17]
```diff
@@ -1,22 +0,0 @@
-// Copyright 2015 Canonical Ltd.
```
Comment (wallyworld) [inline apiserver/facades/client/client/bundles.go]: Don't delete any of this

File: apiserver/facades/client/bundle/mock_test.go (hunk 13 above)
Comment (vinu2003) [reply apiserver/facades/client/bundle/mock_test.go]: done

File: apiserver/facades/client/bundle/bundle.go (hunk 11 above)
Comment (vinu2003) [reply apiserver/facades/client/bundle/bundle.go]: done

File: apiserver/facades/client/bundle/bundle.go [hunk 18]
```diff
@@ -8,47 +8,66 @@ import (
...
 	"github.com/juju/juju/core/devices"
-	"github.com/juju/juju/state"
+	"github.com/juju/juju/permission"
 	"github.com/juju/juju/storage"
 )
 
-// NewFacade provides the required signature for facade registration.
-func NewFacade(_ *state.State, _ facade.Resources, auth facade.Authorizer) (Bundle, error) {
-	return NewBundle(auth)
+// bundleAPI implements the Bundle interface and is the concrete implementation
```
Comment (wallyworld) [inline apiserver/facades/client/bundle/bundle.go]: BundleAPI
(comment needs to match struct name)

File: apiserver/facades/client/bundle/bundle.go [hunk 19]
```diff
@@ -96,3 +115,24 @@ func (b *bundleAPI) GetChanges(args params.BundleChangesParams) (params.BundleCh
...
 }
+
+// ExportBundle exports the current model configuration as bundle.
+func (b *BundleAPI) ExportBundle() (params.StringResult, error) {
+	model, err := b.backend.Export()
+	if err != nil {
+		return params.StringResult{}, errors.Trace(err)
+	}
+
+	if err := b.checkCanRead(model); err != nil {
```
Comment (wallyworld) [inline apiserver/facades/client/bundle/bundle.go]: The auth check is done first before any business logic.
What is passed to the canRead is the model tag taken from the state object. Grab this in the NewFacade() and store as a facade attribute.

File: apiserver/facades/client/bundle/bundle_test.go [hunk 20]
```diff
@@ -222,3 +231,12 @@ func (s *bundleSuite) TestGetChangesBundleEndpointBindingsSuccess(c *gc.C) {
...
 	}
 }
+
+func (s *bundleSuite) TestExportBundleSuccess(c *gc.C) {
+	model, err := s.st.Export()
+	c.Assert(err, jc.ErrorIsNil)
+
+	_, err = description.Serialize(model)
+	c.Check(err, jc.ErrorIsNil)
+	//TODO: need to validate the model.
```
Comment (wallyworld) [inline apiserver/facades/client/bundle/bundle_test.go]: Also check the stub for expected calls

File: apiserver/facades/client/bundle/mock_test.go [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+
+import (
+	"github.com/juju/testing"
+
+	"github.com/juju/description"
+)
+
+type mockState struct {
+	testing.Stub
+	model mockModel
```
Comment (wallyworld) [inline apiserver/facades/client/bundle/mock_test.go]: no need as an attribute here, just construct in export

File: apiserver/facades/client/bundle/mock_test.go [hunk 22]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+func newMockState() *mockState {
+	st := &mockState{
+		model: mockModel{},
+	}
+	return st
+}
+
+type mockModel struct {
```
Comment (wallyworld) [inline apiserver/facades/client/bundle/mock_test.go]: don't need mock model at all

File: apiserver/facades/client/bundle/mock_test.go [hunk 23]
```diff
@@ -0,0 +1,35 @@
...
+	testing.Stub
+	model mockModel
+}
+
+func (m *mockState) Export() (description.Model, error) {
+	m.MethodCall(m, "Export")
+	if err := m.NextErr(); err != nil {
+		return nil, err
+	}
+	return m.model.desc, nil
```
Comment (wallyworld) [inline apiserver/facades/client/bundle/mock_test.go]: the result here should be a model.Description with stuff added that can be checked

File: apiserver/facades/client/bundle/bundle.go (hunk 18 above)
Comment (vinu2003) [reply apiserver/facades/client/bundle/bundle.go]: done

File: apiserver/facades/client/bundle/bundle.go (hunk 19 above)
Comment (vinu2003) [reply apiserver/facades/client/bundle/bundle.go]: Done

File: apiserver/facades/client/bundle/bundle.go [hunk 24]
```diff
@@ -8,47 +8,73 @@ import (
...
 
-// NewFacade provides the required signature for facade registration.
-func NewFacade(_ *state.State, _ facade.Resources, auth facade.Authorizer) (Bundle, error) {
-	return NewBundle(auth)
+// BundleAPI implements the Bundle interface and is the concrete implementation
+// of the API end point.
+type BundleAPI struct {
+	backend    Backend
+	authorizer facade.Authorizer
+	ModelTag   names.ModelTag
```
Comment (wallyworld) [inline apiserver/facades/client/bundle/bundle.go]: no need to export

File: apiserver/facades/client/bundle/bundle_test.go [hunk 25]
```diff
@@ -8,6 +8,8 @@ import (
 	gc "gopkg.in/check.v1"
 	"gopkg.in/juju/names.v2"
```
Comment (wallyworld) [inline apiserver/facades/client/bundle/bundle_test.go]: the imports in this block go together without a newline

File: apiserver/facades/client/bundle/mock_test.go (hunk 22 above)
Comment (vinu2003) [reply apiserver/facades/client/bundle/mock_test.go]: done

File: apiserver/facades/client/bundle/mock_test.go (hunk 21 above)
Comment (vinu2003) [reply apiserver/facades/client/bundle/mock_test.go]: done

File: apiserver/facades/client/bundle/bundle.go [hunk 26]
```diff
@@ -96,3 +122,23 @@ func (b *bundleAPI) GetChanges(args params.BundleChangesParams) (params.BundleCh
 	}
 	return results, nil
 }
+
+// ExportBundle exports the current model configuration as bundle.
+func (b *BundleAPI) ExportBundle() (params.StringResult, error) {
+	if err := b.checkCanRead(); err != nil {
+		return params.StringResult{}, errors.Trace(err)
```
Comment (wallyworld) [inline apiserver/facades/client/bundle/bundle.go]: For errors returned from top level facade functions, use common.ServerError(err)

File: apiserver/facades/client/bundle/bundle_test.go [hunk 27]
```diff
@@ -222,3 +234,16 @@ func (s *bundleSuite) TestGetChangesBundleEndpointBindingsSuccess(c *gc.C) {
 		}
 	}
 }
+
+func (s *bundleSuite) TestExportBundleSuccess(c *gc.C) {
+	_, err := s.auth.HasPermission(permission.ReadAccess, s.modelTag)
```
Comment (wallyworld) [inline apiserver/facades/client/bundle/bundle_test.go]: This is code from the actual implementation - no need to do that here.
If you wanted to test permissions, you'd use a different user name when creating the fake authoriser.

File: apiserver/facades/client/client/bundles.go [hunk 28]
```diff
@@ -14,7 +16,12 @@ import (
 // This call is deprecated, clients should use the GetChanges endpoint on the
 // Bundle facade.
 func (c *Client) GetBundleChanges(args params.BundleChangesParams) (params.BundleChangesResults, error) {
-	bundleAPI, err := bundle.NewBundle(c.api.auth)
+	model, err := c.api.state().Model()
+	if err != nil {
+		return params.BundleChangesResults{}, errors.Trace(err)
```
Comment (wallyworld) [inline apiserver/facades/client/client/bundles.go]: common.ServerErrror() here too

File: apiserver/facades/client/bundle/bundle_test.go [hunk 29]
```diff
@@ -222,3 +234,16 @@ func (s *bundleSuite) TestGetChangesBundleEndpointBindingsSuccess(c *gc.C) {
...
+	c.Assert(err, jc.ErrorIsNil)
+
+	model, err := s.st.Export()
+	c.Assert(err, jc.ErrorIsNil)
+
+	c.Assert(model.Validate(), jc.ErrorIsNil)
+
+	_, err = description.Serialize(model)
```
Comment (wallyworld) [inline apiserver/facades/client/bundle/bundle_test.go]: Need to check actual result value

File: api/modelmanager/modelmanager.go [hunk 30]
```diff
@@ -530,3 +530,17 @@ func (c *Client) UnsetModelDefaults(cloud, region string, keys ...string) error
 	}
 	return result.OneError()
 }
+
+// ExportBundle exports the current model configuration.
+func (c *Client) ExportBundle(tag names.ModelTag) (params.StringResult, error) {
```
Comment (wallyworld) [inline api/modelmanager/modelmanager.go]: It looks like we don't yet have a client for the Bundle facade in the JUju core codebase.
We'll need to create a new api/bundle package and add this method to that.

File: apiserver/facades/client/bundle/bundle.go (hunk 26 above)
Comment (vinu2003) [reply apiserver/facades/client/bundle/bundle.go]: done

File: apiserver/facades/client/bundle/bundle_test.go (hunk 27 above)
Comment (vinu2003) [reply apiserver/facades/client/bundle/bundle_test.go]: done

**Comment by howbazaar** [comment]
@vinu2003 should we be closing this one?

**Comment by anastasiamac** [comment]
Closing as abandoned and moved to a new PR.

**Pull request closed without merging**